File: frontend/src/api/portfolio.ts
export type MarketScenario = "neutral" | "risk_off" | "risk_on";
export type CandidateType = "conservative" | "balanced" | "growth";
export type SuitabilityDecision = "allow" | "downgrade" | "reject" | "education_only";
export type SuitabilityStatus = "pass" | "restrict" | "block";
export type ProductRiskLevel = "r1" | "r2" | "r3" | "r4" | "r5";

export interface SuitabilityCheckItem {
  check_code: string;
  label: string;
  status: SuitabilityStatus;
  observed_value: string;
  rule: string;
  reason: string;
  source_record_ids: string[];
}

export interface SuitabilityGate {
  gate: "family_safety" | "customer" | "product";
  name: string;
  status: SuitabilityStatus;
  decision: SuitabilityDecision;
  effective_risk_limit: ProductRiskLevel | null;
  high_risk_cap: string | null;
  checks: SuitabilityCheckItem[];
  failed_check_codes: string[];
  explanation: string;
}

export interface AllocationLine {
  asset_class: string;
  asset_class_name: string;
  ratio: string;
  amount: string;
  expected_nominal_return: string;
  cvar_loss: string;
  max_drawdown: string;
  liquidity_score: string;
}

export interface ProductMapping {
  asset_class: string;
  asset_class_name: string;
  allocation_ratio: string;
  allocation_amount: string;
  product_id: string | null;
  product_code: string | null;
  product_name: string;
  product_type: string;
  risk_level: ProductRiskLevel | null;
  liquidity_level: string | null;
  decision: SuitabilityDecision;
  reasons: string[];
  is_mock: boolean;
}

export interface RebalanceLine {
  asset_class: string;
  asset_class_name: string;
  current_ratio: string;
  strategic_ratio: string;
  tactical_ratio: string;
  absolute_drift: string;
  relative_drift: string | null;
  trigger: boolean;
  action: string;
}

export interface PortfolioCandidate {
  candidate_type: CandidateType;
  name: string;
  decision: SuitabilityDecision;
  investment_amount: string;
  strategic_allocations: AllocationLine[];
  tactical_allocations: AllocationLine[];
  expected_nominal_return: string;
  expected_real_return: string;
  goal_success_probability: string;
  simulation_method: "deterministic_weighted_scenarios_not_monte_carlo";
  simulation_horizon_months: number;
  simulated_range_low: string;
  simulated_range_base: string;
  simulated_range_high: string;
  extreme_loss_ratio: string;
  extreme_loss_amount: string;
  max_drawdown_estimate: string;
  liquidity_score: string;
  liquidity_description: string;
  annual_fee_rate: string;
  annual_fee_estimate: string;
  applicable_conditions: string[];
  primary_risks: string[];
  why_not_other_candidates: string;
  gates: SuitabilityGate[];
  product_mappings: ProductMapping[];
  optimization: {
    method: "deterministic_grid_search" | "rule_based_fallback";
    status: "optimal" | "fallback" | "infeasible";
    optimizer_version: string;
    random_seed: number;
    grid_step: string;
    evaluated_candidates: number;
    objective_score: string;
    parameter_hash: string;
    parameters: Record<string, unknown>;
    fallback_reason: string | null;
  };
  rebalancing: {
    market_scenario: MarketScenario;
    status: "within_band" | "rebalance_due" | "blocked_by_safety";
    absolute_threshold: string;
    relative_threshold: string;
    maximum_tactical_shift: string;
    next_scheduled_review: string;
    lines: RebalanceLine[];
    explanation: string;
  };
}

export interface MockProduct {
  id: string;
  code: string;
  name: string;
  product_type: string;
  asset_class: string;
  risk_level: ProductRiskLevel;
  term_months: number;
  minimum_holding_months: number;
  liquidity_level: string;
  redemption_rules: string;
  annual_fee_rate: string;
  underlying_assets: string[];
  historical_volatility_min: string;
  historical_volatility_max: string;
  minimum_investment: string;
  suitable_accounts: string[];
  principal_guaranteed: boolean;
  guarantee_basis: string | null;
  guarantee_disclosure: string;
  non_guaranteed_disclosure: string;
  complexity_level: string;
  professional_only: boolean;
  education_only: boolean;
  enabled: boolean;
  is_simulated: boolean;
  terms: Record<string, unknown>;
  catalog_version: string;
  data_date: string;
  source: string;
}

export interface ProductCatalog {
  catalog_code: string;
  catalog_version: string;
  data_date: string;
  source_type: "mock";
  source_summary: string;
  product_count: number;
  products: MockProduct[];
  professional_hedge_lab_enabled: false;
}

export interface PortfolioResponse {
  meta: {
    household_id: string;
    household_code: string;
    analysis_date: string;
    data_as_of: string;
    input_version: string;
    formula_version: string;
    rule_code: string;
    rule_version: string;
    optimizer_version: string;
    catalog_version: string;
    calculation_source: "deterministic_tools";
    currency: string;
    synthetic_data: boolean;
    market_scenario: MarketScenario;
  };
  context: {
    current_growth_assets: string;
    eligible_long_term_amount: string;
    amount_to_restore_safety_layers: string;
    long_term_goal_present_value_gap: string;
    simulation_horizon_months: number;
    annual_new_surplus: string;
    counting_note: string;
  };
  family_safety_gate: SuitabilityGate;
  customer_suitability_gate: SuitabilityGate;
  candidates: PortfolioCandidate[];
  catalog: ProductCatalog;
  education_cards: Array<{ code: string; title: string; summary: string; points: string[] }>;
  professional_hedge_lab: {
    enabled: false;
    mode: "read_only_education";
    title: string;
    reason: string;
    prerequisites: string[];
  };
  counting_note: string;
}

export interface SuitabilityProbeRequest {
  analysis_date?: string;
  investment_amount: string;
  target_horizon_months: number;
  requested_high_risk_ratio: string;
  leverage_ratio?: string;
  concentration_ratio?: string;
  requested_product_codes: string[];
  purpose: "long_term_growth" | "tuition" | "home_purchase" | "retirement" | "professional_hedge";
}

export interface SuitabilityProbeResponse {
  household_id: string;
  decision: SuitabilityDecision;
  gates: SuitabilityGate[];
  failed_check_codes: string[];
  audit_event_id: string;
  rule_version: string;
  calculation_source: "deterministic_tools";
  explanation: string;
}

export interface PersistedPortfolioRun {
  recommendation_id: string;
  portfolio_plan_ids: string[];
  suitability_check_ids: string[];
  household_id: string;
  input_version: string;
  rule_version_id: string;
  rule_version: string;
  candidate_count: number;
  suitability_check_count: number;
  created_at: string;
  calculation_source: "deterministic_tools";
}

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function readJson<T>(path: string, signal?: AbortSignal, init?: RequestInit): Promise<T> {
  const response = await fetch(apiBase + path, {
    ...init,
    signal,
    headers: { Accept: "application/json", ...init?.headers },
  });
  if (!response.ok) throw new Error(`API ${response.status}: ${path}`);
  return (await response.json()) as T;
}

export function fetchPortfolio(
  householdId: string,
  marketScenario: MarketScenario = "neutral",
  signal?: AbortSignal,
): Promise<PortfolioResponse> {
  const query = `?market_scenario=${encodeURIComponent(marketScenario)}`;
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/portfolio${query}`, signal);
}

export function persistPortfolio(
  householdId: string,
  marketScenario: MarketScenario = "neutral",
): Promise<PersistedPortfolioRun> {
  const query = `?market_scenario=${encodeURIComponent(marketScenario)}`;
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/portfolio/runs${query}`,
    undefined,
    { method: "POST" },
  );
}

export function checkPortfolioSuitability(
  householdId: string,
  request: SuitabilityProbeRequest,
): Promise<SuitabilityProbeResponse> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/portfolio/suitability-check`,
    undefined,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    },
  );
}

export function portfolioExportUrl(
  householdId: string,
  marketScenario: MarketScenario = "neutral",
): string {
  return `${apiBase}/api/v1/households/${encodeURIComponent(householdId)}/portfolio/export?market_scenario=${encodeURIComponent(marketScenario)}`;
}

File: backend/app/schemas/portfolio.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.domain.enums import (
    ComplexityLevel,
    LiquidityLevel,
    MarketScenario,
    PortfolioCandidateType,
    ProductRiskLevel,
    SuitabilityDecision,
    SuitabilityGateType,
    SuitabilityStatus,
)
from app.schemas.methodology import DecisionEvidencePackage
from app.schemas.records import Money, Ratio


class ProductCatalogItem(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=160)
    product_type: str = Field(min_length=1, max_length=80)
    asset_class: str = Field(min_length=1, max_length=48)
    risk_level: ProductRiskLevel
    term_months: int = Field(ge=0, le=1200)
    minimum_holding_months: int = Field(ge=0, le=1200)
    liquidity_level: LiquidityLevel
    redemption_rules: str = Field(min_length=1, max_length=800)
    annual_fee_rate: Ratio
    underlying_assets: list[str] = Field(min_length=1)
    historical_volatility_min: Ratio
    historical_volatility_max: Ratio
    minimum_investment: Money
    suitable_accounts: list[str]
    principal_guaranteed: bool
    guarantee_basis: str | None = Field(default=None, max_length=300)
    guarantee_disclosure: str = Field(min_length=1, max_length=800)
    non_guaranteed_disclosure: str = Field(min_length=1, max_length=800)
    complexity_level: ComplexityLevel
    professional_only: bool
    education_only: bool
    enabled: bool
    is_simulated: bool
    terms: dict[str, object] = Field(default_factory=dict)
    account_wrappers: list[str] = Field(default_factory=lambda: ["ordinary"])
    principal_loss_possible: bool | None = None
    legally_principal_guaranteed: bool | None = None
    liquidity_days: int = Field(default=0, ge=0, le=36500)
    lock_up: bool = False
    withdrawable_date: date | None = None
    volatility: Ratio = Decimal("0")
    sale_status: Literal["available", "unavailable", "education_only"] = "available"
    channel: str = "demo_catalog"
    source_reference: str = "Fortune Copilot internal demo catalog"
    snapshot_version: str = "mock-product-snapshot-v1.0.0"

    @model_validator(mode="after")
    def validate_product_boundaries(self) -> ProductCatalogItem:
        if self.legally_principal_guaranteed is None:
            self.legally_principal_guaranteed = self.principal_guaranteed
        if self.principal_loss_possible is None:
            self.principal_loss_possible = not self.legally_principal_guaranteed
        if self.legally_principal_guaranteed != self.principal_guaranteed:
            raise ValueError("兼容保本字段与法律本金保证标签必须一致")
        if self.legally_principal_guaranteed and self.principal_loss_possible:
            raise ValueError("法律本金保证产品不能同时标记本金可损失")
        if self.historical_volatility_min > self.historical_volatility_max:
            raise ValueError("产品波动区间下限不能高于上限")
        if not self.is_simulated:
            raise ValueError("比赛产品目录只允许明确标记为模拟产品")
        if (
            self.product_type
            in {
                "bank_wealth_management",
                "bond_fund",
                "short_bond_fund",
                "collective_fund_trust",
                "participating_life_insurance",
            }
            and self.principal_guaranteed
        ):
            raise ValueError("银行理财、债券基金、信托和分红险不得统一标记为保本")
        if self.education_only and self.enabled:
            raise ValueError("只读教育条目不得开启交易或自动匹配")
        return self


class ProductCatalogFile(BaseModel):
    catalog_code: str
    catalog_version: str
    data_date: date
    source: str
    source_summary: str
    source_system: str
    source_reference: str
    observed_at: date
    effective_at: date
    ingested_at: datetime
    version: str
    data_quality: str
    is_live: bool
    is_demo: bool
    lineage: str
    products: list[ProductCatalogItem] = Field(min_length=14)

    @model_validator(mode="after")
    def validate_unique_codes(self) -> ProductCatalogFile:
        codes = [item.code for item in self.products]
        if len(codes) != len(set(codes)):
            raise ValueError("模拟产品代码必须唯一")
        return self


class ProductOut(ProductCatalogItem):
    model_config = ConfigDict(from_attributes=True)

    id: str
    catalog_version: str
    data_date: date
    source: str
    snapshot_observed_at: date


class ProductCatalogResponse(BaseModel):
    catalog_code: str
    catalog_version: str
    data_date: date
    source_type: Literal["mock"] = "mock"
    source_summary: str
    source_system: str
    source_reference: str
    observed_at: date
    effective_at: date
    ingested_at: datetime
    version: str
    data_quality: str
    is_live: bool
    is_demo: bool
    lineage: str
    product_count: int
    products: list[ProductOut]
    professional_hedge_lab_enabled: bool = False
    snapshot_version: str
    snapshot_observed_at: date
    snapshot_age_days: int
    maximum_age_days: int
    catalog_stale: bool
    executable_recommendations_allowed: bool
    stale_action: Literal["block_executable_allow_education"]


class SuitabilityCheckItem(BaseModel):
    check_code: str
    label: str
    status: SuitabilityStatus
    observed_value: str
    rule: str
    reason: str
    source_record_ids: list[str] = Field(default_factory=list)


class SuitabilityGateResult(BaseModel):
    gate: SuitabilityGateType
    name: str
    status: SuitabilityStatus
    decision: SuitabilityDecision
    effective_risk_limit: ProductRiskLevel | None = None
    high_risk_cap: Ratio | None = None
    checks: list[SuitabilityCheckItem]
    failed_check_codes: list[str]
    explanation: str


class AllocationLine(BaseModel):
    asset_class: str
    asset_class_name: str
    ratio: Ratio
    amount: Money
    expected_nominal_return: Ratio
    cvar_loss: Ratio
    max_drawdown: Ratio
    liquidity_score: Ratio


class ProductMapping(BaseModel):
    asset_class: str
    asset_class_name: str
    allocation_ratio: Ratio
    allocation_amount: Money
    product_id: str | None
    product_code: str | None
    product_name: str
    product_type: str
    risk_level: ProductRiskLevel | None
    liquidity_level: LiquidityLevel | None
    decision: SuitabilityDecision
    reasons: list[str]
    is_mock: bool = True


class OptimizationDiagnostics(BaseModel):
    method: Literal["deterministic_grid_search", "rule_based_fallback"]
    status: Literal["optimal", "fallback", "infeasible"]
    optimizer_version: str
    random_seed: int
    grid_step: Ratio
    evaluated_candidates: int
    objective_score: Decimal
    parameter_hash: str
    parameters: dict[str, object]
    fallback_reason: str | None = None


class RebalanceLine(BaseModel):
    asset_class: str
    asset_class_name: str
    current_ratio: Ratio
    strategic_ratio: Ratio
    tactical_ratio: Ratio
    absolute_drift: Ratio
    relative_drift: Decimal | None = Field(default=None, ge=0)
    trigger: bool
    action: str


class RebalancePlan(BaseModel):
    market_scenario: MarketScenario
    status: Literal["within_band", "rebalance_due", "blocked_by_safety"]
    absolute_threshold: Ratio
    relative_threshold: Ratio
    maximum_tactical_shift: Ratio
    next_scheduled_review: date
    lines: list[RebalanceLine]
    explanation: str


class PortfolioCandidate(BaseModel):
    candidate_type: PortfolioCandidateType
    name: str
    decision: SuitabilityDecision
    investment_amount: Money
    strategic_allocations: list[AllocationLine]
    tactical_allocations: list[AllocationLine]
    expected_nominal_return: Ratio
    expected_real_return: Decimal
    goal_success_probability: Ratio
    simulation_method: Literal["deterministic_weighted_scenarios_not_monte_carlo"]
    simulation_horizon_months: int
    simulated_range_low: Money
    simulated_range_base: Money
    simulated_range_high: Money
    extreme_loss_ratio: Ratio
    extreme_loss_amount: Money
    max_drawdown_estimate: Ratio
    liquidity_score: Ratio
    liquidity_description: str
    annual_fee_rate: Ratio
    annual_fee_estimate: Money
    responsibility_breach_probability: Ratio
    purchasing_power_success_probability: Ratio
    liability_coverage: Ratio
    liquidity_shortfall: Ratio
    concentration: Ratio
    real_return_after_fee: Decimal
    applicable_conditions: list[str]
    primary_risks: list[str]
    why_not_other_candidates: str
    gates: list[SuitabilityGateResult]
    product_mappings: list[ProductMapping]
    optimization: OptimizationDiagnostics
    rebalancing: RebalancePlan


class PortfolioContext(BaseModel):
    current_growth_assets: Money
    eligible_long_term_amount: Money
    amount_to_restore_safety_layers: Money
    long_term_goal_present_value_gap: Money
    simulation_horizon_months: int
    annual_new_surplus: Money
    purchasing_power_hurdle: Ratio
    single_equity_amount: Money
    largest_single_security_ratio: Ratio
    single_equity_hhi: Ratio
    security_concentration_treatment: Literal["satellite_only"] = "satellite_only"
    counting_note: str


class PortfolioMeta(BaseModel):
    household_id: str
    household_code: str
    analysis_date: date
    data_as_of: date
    input_version: str
    formula_version: str
    rule_code: str
    rule_version: str
    optimizer_version: str
    catalog_version: str
    calculation_source: Literal["deterministic_tools"] = "deterministic_tools"
    currency: str = "CNY"
    synthetic_data: bool
    market_scenario: MarketScenario
    methodology_version: str
    market_regime_version: str
    product_snapshot_version: str


class EducationCard(BaseModel):
    code: str
    title: str
    summary: str
    points: list[str]


class HedgeLabBoundary(BaseModel):
    enabled: Literal[False] = False
    mode: Literal["read_only_education"] = "read_only_education"
    title: str
    reason: str
    prerequisites: list[str]


class PortfolioResponse(BaseModel):
    meta: PortfolioMeta
    context: PortfolioContext
    family_safety_gate: SuitabilityGateResult
    customer_suitability_gate: SuitabilityGateResult
    candidates: list[PortfolioCandidate] = Field(min_length=3, max_length=3)
    catalog: ProductCatalogResponse
    education_cards: list[EducationCard]
    professional_hedge_lab: HedgeLabBoundary
    decision_evidence: DecisionEvidencePackage
    counting_note: str


class SuitabilityProbeRequest(BaseModel):
    analysis_date: date | None = None
    investment_amount: Money
    target_horizon_months: int = Field(ge=1, le=1200)
    requested_high_risk_ratio: Ratio
    leverage_ratio: Ratio = Decimal("0")
    concentration_ratio: Ratio = Decimal("0")
    requested_product_codes: list[str] = Field(min_length=1, max_length=20)
    purpose: Literal[
        "long_term_growth",
        "tuition",
        "home_purchase",
        "retirement",
        "professional_hedge",
    ]


class SuitabilityProbeResponse(BaseModel):
    household_id: str
    decision: SuitabilityDecision
    gates: list[SuitabilityGateResult]
    failed_check_codes: list[str]
    audit_event_id: str
    rule_version: str
    calculation_source: Literal["deterministic_tools"] = "deterministic_tools"
    explanation: str


class PersistedPortfolioRun(BaseModel):
    recommendation_id: str
    portfolio_plan_ids: list[str]
    suitability_check_ids: list[str]
    household_id: str
    input_version: str
    rule_version_id: str
    rule_version: str
    candidate_count: int
    suitability_check_count: int
    created_at: datetime
    calculation_source: Literal["deterministic_tools"] = "deterministic_tools"
